src/query/plan/hint_provider.hpp

```diff
@@ -0,0 +1,266 @@
+    return false;
+  }
+
+  bool PostVisit(Foreach &) override { return true; }
+
+  bool PreVisit(EvaluatePatternFilter &op) override { return true; }
+
+  bool PostVisit(EvaluatePatternFilter & /*op*/) override { return true; }
+
+  bool PreVisit(Apply &op) override {
+    op.input()->Accept(*this);
+    op.subquery_->Accept(*this);
+    return false;
+  }
+
+  bool PostVisit(Apply & /*op*/) override { return true; }
+
+  bool PreVisit(LoadCsv &op) override { return true; }
+
+  bool PostVisit(LoadCsv & /*op*/) override { return true; }
+
+ private:
+  const SymbolTable &symbol_table_;
+  std::vector<std::string> hints_;
+
+  bool DefaultPreVisit() override { throw utils::NotYetImplemented("providing plan hints"); }
+
+  void HintIndexUsage(Filter &op) {
+    if (auto *maybe_scan_operator = dynamic_cast<ScanAll *>(op.input().get()); !maybe_scan_operator) {
+      return;
+    }
+
+    auto const scan_symbol = dynamic_cast<ScanAll *>(op.input().get())->output_symbol_;
+    auto const scan_type = op.input()->GetTypeInfo();
+
+    Filters filters;
+    filters.CollectFilterExpression(op.expression_, symbol_table_);
+    const std::string filtered_labels = ExtractAndJoin(filters.FilteredLabels(scan_symbol),
+                                                       [](const auto &item) { return fmt::format(":{0}", item.name); });
+    const std::string filtered_properties =
+        ExtractAndJoin(filters.FilteredProperties(scan_symbol), [](const auto &item) { return item.name; });
+
+    if (filtered_labels.empty() && filtered_properties.empty()) {
+      return;
+    }
+
+    if (scan_type == ScanAll::kType) {
+      if (!filtered_labels.empty() && !filtered_properties.empty()) {
+        hints_.push_back(
+            fmt::format("Sequential scan will be used on symbol `{0}` although there is a filter on labels {1} and "
+                        "properties {2}. Consider "
+                        "creating a label property index.",
+                        scan_symbol.name(), filtered_labels, filtered_properties));
+        return;
+      }
+
+      if (!filtered_labels.empty()) {
+        hints_.push_back(fmt::format(
+            "Sequential scan will be used on symbol `{0}` although there is a filter on labels {1}. Consider "
+            "creating a label index.",
+            scan_symbol.name(), filtered_labels));
+        return;
+      }
+      return;
+    }
+
+    if (scan_type == ScanAllByLabel::kType && !filtered_properties.empty()) {
+      hints_.push_back(fmt::format(
+          "Label index will be used on symbol `{0}` although there is also a filter on properties {1}. Consider "
+          "creating a label property index.",
+          scan_symbol.name(), filtered_properties));
+      return;
+    }
+  }
+
+  std::string ExtractAndJoin(auto &&collection, auto &&projection) {
+    auto elements = collection | ranges::views::transform(projection);
+    return boost::algorithm::join(elements, ", ");
```

Comment:
you have ours utils::Join

Reply:
i tried to switch but it seems the elements are not a plain vector of strings but a view. so I'll keep it because boost manages to work with it